optuna/samplers/_brute_force.py

```diff
@@ -0,0 +1,115 @@
+import copy
+import decimal
+from typing import Any
+from typing import Dict
+from typing import Optional
+from typing import Sequence
+
+import numpy as np
+
+from optuna.distributions import BaseDistribution
+from optuna.distributions import CategoricalDistribution
+from optuna.distributions import FloatDistribution
+from optuna.distributions import IntDistribution
+from optuna.samplers import BaseSampler
+from optuna.study import Study
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+class BruteForceSampler(BaseSampler):
+    """Sampler using brute force.
+
+    This sampler performs exhaustive search on the defined search space.
+
+    Note:
+        The defined search space must be finite. Therefore, when using
+        :class:`~optuna.distributions.FloatDistibution`, `step=None` is not allowed.
+
+    Note:
+        This sampler should not be used in combination with other samplers. For example,
+        in distributed optimization, if :class:`~optuna.samplers.BruteForceSampler` and another
+        sampler run optimizations simultaneously, complete exhaustive search may not be performed.
+
+    Note:
+        The objective function must be fixed during the search. Otherwise, sampler may fail to
+        try the entire search space.
+
+    Args:
+        seed:
+            A seed to fix the order of trials as the search order randomly shuffled. Please note
+            that it is not recommended using this option in distributed optimization settings since
+            this option cannot ensure the order of trials and may increase the number of duplicate
+            suggestions during distributed optimization.
+    """
+
+    def __init__(self, seed: Optional[int] = None) -> None:
+        self._rng = np.random.RandomState(seed)
+
+    def infer_relative_search_space(
+        self, study: Study, trial: FrozenTrial
+    ) -> Dict[str, BaseDistribution]:
+        return {}
+
+    def sample_relative(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> Dict[str, Any]:
+        return {}
+
+    def sample_independent(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+        candidates = _enumerate_candidates(param_distribution)
+        if len(candidates) == 0:
+            raise ValueError("empty candidates")
+
+        self._rng.shuffle(candidates)
+
+        for value in candidates[1:]:
+            params = copy.deepcopy(trial.params)
```

Comment:
```suggestion
            params = trial.params.copy()
```
I think a shallow copy is enough.